File: golang/stack/astreroid_collision.go
package main

func asteroidCollision(asteroids []int) []int {
	stack := []int{}

	for _, asteroid := range asteroids {
		for len(stack) > 0 && asteroid < 0 && stack[len(stack)-1] > 0 {
			if stack[len(stack)-1] > -asteroid {
				asteroid = 0

				break
			} else if stack[len(stack)-1] < -asteroid {
				stack = stack[:len(stack)-1]
			} else {
				stack = stack[:len(stack)-1]

				asteroid = 0

				break
			}
		}

		if asteroid != 0 {
			stack = append(stack, asteroid)
		}
	}

	return stack
}

File: java/src/main/java/com/sanedge/interview/stack/AstreroidCollision.java
package com.sanedge.interview.stack;

import java.util.Stack;

public class AstreroidCollision {
  public int[] asteroidCollision(int[] asteroids) {
    Stack<Integer> stack = new Stack<>();

    for (int asteroid : asteroids) {
      while (!stack.isEmpty() && asteroid < 0 && stack.peek() > 0) {
        if (stack.peek() > -asteroid) {
          asteroid = 0;
          break;
        } else if (stack.peek() < -asteroid) {
          stack.pop();
        } else {
          stack.pop();
          asteroid = 0;
          break;
        }
      }

      if (asteroid != 0) {
        stack.push(asteroid);
      }
    }

    int[] result = new int[stack.size()];
    for (int i = 0; i < result.length; i++) {
      result[i] = stack.get(i);
    }

    return result;
  }

}
